### PR TITLE
[Refactor] Task and Tune API

### PR DESCRIPTION
**Pull request by KKIEEK**
## Motivation

To build runnable task from cfg

### TODO
- [x] Enhance `work_dir` (experiment name is determined by `work_dir`)
- [x] Refactor `log_analysis`
- [ ] Add test code for `Tuner` class
- [ ] Pass test code
- [ ] Final test

## Modification

Please briefly describe what modification is made in this PR.

## Expected behavior
```
work_dirs
  |- mmdet_asynchb_nevergrad_pso (config name)
    |- 221221_113600 (experiment name)
      |- best_trial
        |- work_dirs (log of best trial)
        |- best_trial.log
      |- DataParallelTrainer_738f915b_1_data_samples_per_gpu=5,model=yolo_x_8x8,optimizer=adam ... (trial_name)
        |- rank_0 (log dir for task process)
          |- work_dirs
            |- faster_rcnn_r50_fpn_1x_coco
              |- faster_rcnn_r50_fpn_1x_coco.py
              |- latest.pth
        |- rank_1
        |- params.json
        |- params.pkl 
        |- progress.csv
        |- result.json
      |- DataParallelTrainer_7caa5acf_2_data_samples_per_gpu=6,model=faster_rcnn_swin_s_p4_w7_fpn,optimizer=adamw ...
        |- (trial directory name is determined by trial_name_creator and trial_dirname_creator)
      |- some.json
      |- some.pkl
```

## Result
```
work_dirs/mmdet_asynchb_nevergrad_pso/20221223_070902
├── 1_data_samples_per_gpu=7,model=yolo_x_8x8,optimizer=adamw
│   ├── error.pkl
│   ├── error.txt
│   ├── events.out.tfevents.1671779346.jhsong-test-hpo-6667d579dc-nvhjg
│   ├── params.json
│   ├── params.pkl
│   ├── rank_0
│   │   └── work_dirs
│   │       └── 2005c6e6
│   │           ├── 2005c6e6.py
│   │           ├── 20221223_070919.log
│   │           └── 20221223_070919.log.json
│   └── result.json
├── 2_data_samples_per_gpu=3,model=tood_r101_dcnv2,optimizer=adamw
│   ├── checkpoint_-00001
│   ├── events.out.tfevents.1671779388.jhsong-test-hpo-6667d579dc-nvhjg
│   ├── params.json
│   ├── params.pkl
│   ├── progress.csv
│   ├── rank_0
│   │   └── work_dirs
│   │       └── 60d57f42
│   │           ├── 20221223_071001.log
│   │           ├── 20221223_071001.log.json
│   │           └── 60d57f42.py
│   └── result.json
├── 3_data_samples_per_gpu=6,model=yolo_x_8x8,optimizer=rms
│   ├── error.pkl
│   ├── error.txt
│   ├── events.out.tfevents.1671779444.jhsong-test-hpo-6667d579dc-nvhjg
│   ├── params.json
│   ├── params.pkl
│   ├── rank_0
│   │   └── work_dirs
│   │       └── 777781b3
│   │           ├── 20221223_071057.log
│   │           ├── 20221223_071057.log.json
│   │           └── 777781b3.py
│   └── result.json
├── 4_data_samples_per_gpu=3,model=tood_r101_dcnv2,optimizer=sgd
│   ├── checkpoint_-00001
│   ├── events.out.tfevents.1671779486.jhsong-test-hpo-6667d579dc-nvhjg
│   ├── params.json
│   ├── params.pkl
│   ├── progress.csv
│   ├── rank_0
│   │   └── work_dirs
│   │       └── b42854ef
│   │           ├── 20221223_071139.log
│   │           ├── 20221223_071139.log.json
│   │           └── b42854ef.py
│   └── result.json
├── 5_data_samples_per_gpu=2,model=faster_rcnn_x101_64x4d_fpn,optimizer=rms
│   ├── checkpoint_-00001
│   ├── events.out.tfevents.1671779532.jhsong-test-hpo-6667d579dc-nvhjg
│   ├── params.json
│   ├── params.pkl
│   ├── progress.csv
│   ├── rank_0
│   │   └── work_dirs
│   │       └── 463aa231
│   │           ├── 20221223_071225.log
│   │           ├── 20221223_071225.log.json
│   │           └── 463aa231.py
│   └── result.json
├── 6_data_samples_per_gpu=3,model=faster_rcnn_swin_s_p4_w7_fpn,optimizer=sgd
│   ├── checkpoint_-00001
│   ├── events.out.tfevents.1671779574.jhsong-test-hpo-6667d579dc-nvhjg
│   ├── params.json
│   ├── params.pkl
│   ├── progress.csv
│   ├── rank_0
│   │   └── work_dirs
│   │       └── 037ff5be
│   │           ├── 037ff5be.py
│   │           ├── 20221223_071308.log
│   │           └── 20221223_071308.log.json
│   └── result.json
├── 7_data_samples_per_gpu=7,model=yolo_x_8x8,optimizer=adam
│   ├── error.pkl
│   ├── error.txt
│   ├── events.out.tfevents.1671779618.jhsong-test-hpo-6667d579dc-nvhjg
│   ├── params.json
│   ├── params.pkl
│   ├── rank_0
│   │   └── work_dirs
│   │       └── 7b5d2242
│   │           ├── 20221223_071351.log
│   │           ├── 20221223_071351.log.json
│   │           └── 7b5d2242.py
│   └── result.json
├── 8_data_samples_per_gpu=5,model=faster_rcnn_x101_64x4d_fpn,optimizer=adam
│   ├── checkpoint_-00001
│   ├── events.out.tfevents.1671779660.jhsong-test-hpo-6667d579dc-nvhjg
│   ├── params.json
│   ├── params.pkl
│   ├── progress.csv
│   ├── rank_0
│   │   └── work_dirs
│   │       └── 61b2f6a0
│   │           ├── 20221223_071433.log
│   │           ├── 20221223_071433.log.json
│   │           └── 61b2f6a0.py
│   └── result.json
├── best_trial
│   ├── best_trial.log
│   └── log
│       ├── checkpoint_-00001
│       ├── events.out.tfevents.1671779388.jhsong-test-hpo-6667d579dc-nvhjg
│       ├── params.json
│       ├── params.pkl
│       ├── progress.csv
│       ├── rank_0
│       │   └── work_dirs
│       │       └── 60d57f42
│       │           ├── 20221223_071001.log
│       │           ├── 20221223_071001.log.json
│       │           └── 60d57f42.py
│       └── result.json
├── experiment_state-2022-12-23_07-09-06.json
├── mmdet_asynchb_nevergrad_pso.py
├── search_gen_state-2022-12-23_07-09-06.json
├── searcher-state-2022-12-23_07-09-06.pkl
├── trainable.pkl
└── tuner.pkl
```

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: siatune/tune/tuner.py [hunk 1]
```diff
@@ -34,11 +36,12 @@ class Tuner:
             Refer to :class:`ray.tune.callback.Callback` for more info.
         resume (str, optional): The experiment path to resume.
             Default to None.
+        cfg (dict, optional) Full config. Default to None.
     """
 
     def __init__(
         self,
-        trainable: Callable[[dict], Any],
+        task: dict,
```
Comment (nijkah) [inline siatune/tune/tuner.py:44]: How about `task_cfg` as in the line 44?

Comment (nijkah) [reply siatune/tune/tuner.py:44]: To my current thought, all of `Callable` can be given by defined `Task`s.
So I don't feel the necessity to support `Callable` in the current status.

Comment (nijkah) [inline siatune/tune/tuner.py:44]:
```suggestion
        task_cfg: Config,
```

**Review by yhna940 — COMMENTED** [review]
GOOD 💡

**Comment by KKIEEK** [comment]
```python3
# at tune/config.py
class _CustomTorchBackend(_TorchBackend):
    ...

        def set_env_vars(addr, port, rank, world_size):
            # fix here
            os.environ[TUNE_ORIG_WORKING_DIR] = os.getcwd()

            os.environ['MASTER_ADDR'] = addr
            os.environ['MASTER_PORT'] = str(port)
            os.environ['RANK'] = str(rank)
            os.environ['LOCAL_RANK'] = str(rank)
            os.environ['WORLD_SIZE'] = str(world_size)
```

```python3
# at core/rewriters/rel2abs.py
import os

from .base import BaseRewriter
from .builder import REWRITERS

@REWRITERS.register_module()
class Rel2Abs(BaseRewriter):

    def __call__(self, context: dict) -> dict:
        os.chdir(os.environ[TUNE_ORIG_WORKING_DIR])
        return context
```

File: siatune/tune/tuner.py (hunk 1 above)
Comment (KKIEEK) [reply siatune/tune/tuner.py:49]: I won't fix it.

**Review by nijkah — APPROVED** [review]

**Pull request merged**